test/src/test/java/hudson/model/CauseTest.java

```diff
@@ -74,9 +74,6 @@ public class CauseTest {
         assertTrue("keeps full history:\n" + buildXml, buildXml.contains("<upstreamBuild>1</upstreamBuild>"));
         buildXml = new XmlFile(Run.XSTREAM, new File(last.getRootDir(), "build.xml")).asString();
         assertFalse("too big:\n" + buildXml, buildXml.contains("<upstreamBuild>1</upstreamBuild>"));
-        for (QueueTaskFuture<FreeStyleBuild> future : futures) {
-            j.assertBuildStatusSuccess(j.waitForCompletion(future.waitForStart()));
```

Comment:
By the way we do require that all scheduled or started builds are completed (or canceled before they start) before the end of the test; otherwise there can be a mess cleaning up after the test’s temp dir. I think you are doing that here.

Reply:
line 105 handles all of the builds in this test as well as the extra handling added by the looks of it

Reply:
For `deeplyNestedCauses`, every scheduled build is now passed to `j.waitForCompletion` immediately so this snippet is no longer necessary.

For `broadlyNestedCauses`, as @timja mentioned, line 105 will clean up any lingering scheduled builds, however every single build was either:
1. scheduled with a quiet period of 1 and recorded in `futures`
2. scheduled with a quiet period of 0 and immediately awaited for completion

Additionally, each case of (1) is followed by an instance of (2) so the aforementioned cleanup on line 105 should never actually do anything. Though some builds were scheduled without immediately awaiting, there is always a subsequent scheduling of that same job that does immediately await.